Scan an assembly in order to create a DataContractContentSerializer

moodmosaic opened the pull request:
Resolves #100.

moodmosaic committed:
Added DataContractContentSerializer.Scan.

moodmosaic committed:
Returned a DataContractContentSerializer instance.

moodmosaic committed:
Added a test case.

moodmosaic committed:
Separated a Guard Clause with a blank line.

moodmosaic committed:
Documented DataContractContentSerializer.Scan.

moodmosaic committed:
Replaced an assertion with the most strict one.

moodmosaic committed:
Replaced an assertion with the most strict one.

ploeh commented on AtomEventStore.UnitTests/DataContractContentSerializerTests.cs:
```diff
@@ -113,5 +113,61 @@ where t.IsDefined(a.GetType(), inherit: false)
+        }
+
+        [Theory, AutoAtomData]
+        public void ScanCorrectlySerializesAttributedClassInstance(
```
Wouldn't it be simpler just to verify that the Scan method returns an object with well-known behaviour?

moodmosaic replied:
Assuming the test did only this:

``` c#
Assert.IsType<DataContractContentSerializer>(
    DataContractContentSerializer.Scan(assemblyToScanForEvents));
```

Would this false implementation pass the test?

``` c#
public static IContentSerializer Scan(Assembly assemblyToScanForEvents)
{
    if (assemblyToScanForEvents == null)
        throw new ArgumentNullException("assemblyToScanForEvents");

    var resolver =
        DataContractContentSerializer.CreateTypeResolver(
            typeof(Version).Assembly); // Instead of typeof(assemblyToScanForEvents).Assembly
    return new DataContractContentSerializer(resolver);
}
```

ploeh replied:
OK. I see. Thank you for the explanation.

moodmosaic replied:
Perhaps we could get rid of the test-code duplication, e.g. invoking the rule of three, eventually.

ploeh commented on AtomEventStore.UnitTests/DataContractContentSerializerTests.cs:
```diff
@@ -113,5 +113,61 @@ where t.IsDefined(a.GetType(), inherit: false)
+            DataContractTestEventX dctex)
+        {
+            var assemblyToScanForEvents = dctex.GetType().Assembly;
+            var sut =
```
Is this really the System Under Test?

moodmosaic replied:
Essentially, SUT is the return value of `Scan`. If we separate the Guard Assertion it looks like:

``` c#
var sut = DataContractContentSerializer.Scan(assemblyToScanForEvents);
Assert.IsType<DataContractContentSerializer>(sut);
```

ploeh replied:
Why is SUT the _return_ value of `Scan`? What are we testing here?

moodmosaic replied:
Scan is acting like a Factory Method, isn't it?

ploeh replied:
No, [Factory Method](http://en.wikipedia.org/wiki/Factory_method_pattern) is a design pattern that relies on inheritance to decouple creation from usage. It's not related to the `Scan` method.

moodmosaic replied:
Oh :$ I see... Thank you for clarifying this.

Unless I'm missing something, `Scan` returns an _IContentSerializer_ which is actually a `DataContractContentSerializer`.

How can we verify that `Scan` _returns correct result_?
- Verify that it's actually a `DataContractContentSerializer`.
- Verify that it's created with the expected `ITypeResolver`.

ploeh replied:
Let's take it in a different direction. I find the test difficult to read because, by looking at it, it's not clear to me _what_ is being tested.

Normally, I look for a `sut` variable, and if I find one, I look for a [single, isolated line of code that exercises it](http://blog.ploeh.dk/2013/06/24/a-heuristic-for-formatting-code-according-to-the-aaa-pattern). However, here, the `sut` variable is only created _after_ the behaviour that the test is exercising has been invoked, and I find that confusing.

The System Under Test is the system being _exercised_ by the test. What is being exercised here?

moodmosaic replied:
> Let's take it in a different direction. I find the test difficult to read because, by looking at it, it's not clear to me _what_ is being tested.

Agreed.

> What is being exercised here?

We're exercising the behavior of the value returned by the `Scan` function. We do that because `Scan` internally creates and returns a new instance, and we want to make sure that this new instance is created _as expected_.

> However, here, the `sut` variable is only created _after_ the behaviour that the test is exercising has been invoked, and I find that confusing.

Agreed. I wish we could do Structural Inspection instead...

ploeh replied:
> We're exercising the behavior of the value returned by the Scan function.

Ah, I see. That's where we view things differently :)

The way I see it, this test is part of a TDD process. The _goal_ of this particular process is to produce the `Scan` method. In my opinion, then, the `Scan` method is the System Under Test.

The rest is just verification.

> I wish we could do Structural Inspection instead.

Yes; and we _could_ have, if we'd created two new implementations of `ITypeResolver`, instead of reusing `TypeResolutionTable` within the `CreateTypeResolver` methods... :)

moodmosaic replied:
Even if we do that, `Scan` accepts a single argument and `DataContractContentSerializer` doesn't expose an `ITypeResolver` inspection property.

ploeh replied:
Couldn't it, though?

moodmosaic replied:
Sure, why not (now that we need it).

ploeh replied:
To be clear, I only meant the following as a throw-away comment:

> Yes; and we could have, if we'd created two new implementations of ITypeResolver, instead of reusing TypeResolutionTable within the CreateTypeResolver methods.

We don't _need_ to do that - in fact, now I'm sorry I wrote that at all, because it seems to derail us a little from the actual issue at hand.

moodmosaic replied:
Hm... AFAICT, we've done [something similar](https://github.com/GreanTech/AtomEventStore/blob/bdee64d1d7f5cd8ea03bafdfcc6927ef7aa7a736/AtomEventStore.UnitTests/LifoEventsTests.cs#L136-141) with `FifoEvents` and `LifoEvents` `Reverse` method.

Couldn't we follow the same direction here?

ploeh replied:
Perhaps... either that, or make these two tests more readable.

moodmosaic replied:
> or make these two tests more readable

I'd like to try this first, as it doesn't require changes to the SUT.

How about doing _something like_ this:

``` c#
[Theory, AutoData]
public void ScanCorrectlySerializesAttributedClassInstance(
    DataContractTestEventX @event)
{
    var sut = DataContractContentSerializer.Scan(@event.GetType().Assembly);

    var actual = @event.AsSerializedString(sut);

    var expected = XDocument.Parse(
        "<test-event-x xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://grean.rocks/dc\">" +
        "  <number>" + @event.Number + "</number>" +
        "  <text>" + @event.Text + "</text>" +
        "</test-event-x>");
    Assert.Equal(expected, XDocument.Parse(actual), new XNodeEqualityComparer());
}
```

The `AsSerializedString` is a [Feature Envy](http://c2.com/cgi/wiki?FeatureEnvySmell) defined as:

``` c#
internal static class DataContractTestEventEnvy
{
    internal static string AsSerializedString(
        this IDataContractTestEvent @event,
        IContentSerializer serializer)
    {
        var sb = new StringBuilder();
        using (var w = XmlWriter.Create(sb))
        {
            serializer.Serialize(w, @event);
            w.Flush();
        }
        return sb.ToString();
    }
}
```

ploeh replied:
That looks more readable, but I don't follow the variable naming. Why is `sut` and `actual` named like that?

It might help to identify what's what if you try to format the test according to [this heuristic](http://blog.ploeh.dk/2013/06/24/a-heuristic-for-formatting-code-according-to-the-aaa-pattern).

moodmosaic replied:
> Why is sut and actual named like that?

`sut` is the result produced by the `Scan` method and `actual` is the serialized XML produced by the `sut`.

IMHO it's not enough to assume that `sut` is just the result of `Scan`—We have to do some sort of verification, otherwise one can easily use the [Devil's Advocate technique](http://www.pluralsight.com/courses/outside-in-tdd) inside the `Scan` method...

ploeh replied:
What does the `sut` name _mean_?

moodmosaic replied:
Isn't the result of `Scan` the System Under Test? (That's what the `sut` name means.)

ploeh replied:
Which production code did you add in order to make these three tests pass?

moodmosaic replied:
Isn't the answer already [there](https://github.com/GreanTech/AtomEventStore/pull/102/files#r19254947)? (To avoid running into a [paradox](http://en.wikipedia.org/wiki/I_know_that_I_know_nothing).)

ploeh replied:
Apparently not...

moodmosaic replied:
I started by adding [this](https://github.com/moodmosaic/AtomEventStore/commit/12d4a71c94b5499ddd0f9f92d974cb533d9c20dc) commit and then I added [this](https://github.com/moodmosaic/AtomEventStore/commit/003c6f87a4076918bdefbcf1aa7937068596ef52) commit, in order to return the correct `IContentSerializer` result.

ploeh replied:
So, _what_ are those tests testing?

moodmosaic replied:
Almost exclusively the _behavior_ of the returned `IContentSerializer` from the `Scan` method.

ploeh replied:
The `Scan` method returns an instance of `DataContractContentSerializer`. Did the behaviour of that class change as a result of these tests being added?

moodmosaic replied:
No, it doesn't. However, _any_ instance of `DataContractContentSerializer` is allowed to be returned from `Scan`?

ploeh replied:
If the behaviour of `DataContractContentSerializer` didn't change in reaction to those tests, is it then the System Under Test?

moodmosaic replied:
Probably not. The question is, how can we know that `Scan` returns the correct result?

ploeh replied:
I think your [suggested verification](https://github.com/GreanTech/AtomEventStore/pull/102/files#r19732669) is fine - I just don't find the variable names or the white space helpful.

What's the System Under Test?

Hint: there's a reason it's called _System_ Under Test, and not Object Under Test. Just like _unit_ doesn't mean _object_.

moodmosaic replied:
_Oh_, I see now... Yes, agreed. How about renaming `sut` to `serializer`?

ploeh replied:
How about renaming `sut` to `actual`?

moodmosaic replied:
Renaming `sut` to `actual` means `actual` must first be renamed to something else, e.g. `result`:

``` c#
[Theory, AutoData]
public void ScanCorrectlySerializesAttributedClassInstance(
    DataContractTestEventX @event)
{
    var actual = DataContractContentSerializer.Scan(@event.GetType().Assembly);
    var result = @event.AsSerializedString(actual);
    var expected = XDocument.Parse(
        "<test-event-x xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://grean.rocks/dc\">" +
        "  <number>" + @event.Number + "</number>" +
        "  <text>" + @event.Text + "</text>" +
        "</test-event-x>");
    Assert.Equal(expected, XDocument.Parse(result), new XNodeEqualityComparer());
}
```

Or we could get rid of `result` entirely:

``` c#
[Theory, AutoData]
public void ScanCorrectlySerializesAttributedClassInstance(
    DataContractTestEventX @event)
{
    var actual = DataContractContentSerializer.Scan(@event.GetType().Assembly);
    var expected = XDocument.Parse(
        "<test-event-x xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://grean.rocks/dc\">" +
        "  <number>" + @event.Number + "</number>" +
        "  <text>" + @event.Text + "</text>" +
        "</test-event-x>");
    Assert.Equal(
        expected,
        XDocument.Parse(@event.AsSerializedString(actual)),
        new XNodeEqualityComparer());
}
```

ploeh replied:
Right. So, to make it explicit, the System Under Test is the `Scan` method. That's the method you added after having written the tests. There's no _sut_ object, because the SUT isn't an object, but a static method.

(When unit testing F# code, I rarely have a `sut` value, because the SUTs are functions, and functions are rarely values (not until they become closures, that is).)

Now that you know what the `actual` result is, then where are the Arrange, Act, and Assert phases in this test?

moodmosaic replied:
> When unit testing F# code, I rarely have a sut value, because the SUTs are functions, and functions are rarely values

I think that's a great and valuable observation! (May also fit into 140 chars.)

---

According to the Arrange, Act, and Assert phases, the test could be written as:

``` c#
[Theory, AutoData]
public void ScanCorrectlySerializesAttributedClassInstance(
    DataContractTestEventX @event)
{
    // Arrange
    var expected = XDocument.Parse(
        "<test-event-x xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://grean.rocks/dc\">" +
        "  <number>" + @event.Number + "</number>" +
        "  <text>" + @event.Text + "</text>" +
        "</test-event-x>");
    var assembly = @event.GetType().Assembly;
    // Act
    var actual = DataContractContentSerializer.Scan(assembly);
    var result = XDocument.Parse(@event.AsSerializedString(actual));
    // Assert
    Assert.Equal(expected, result, new XNodeEqualityComparer());
}
```

ploeh replied:
Is there any reason to introduce `expected` in the _Arrange_ phase?

Why would you say that invoking `XDocument.Parse` is part of the _Act_ phase?

(Hint: one of the phases is degenerate)

moodmosaic replied:
When the `expected` value requires some setup (in this example, string formatting, concatenation, etc) I usually put it in the _Arrange_ phase. That might be a mistake on my part though, as I didn't look-up that in Gerard Meszaros's xUnit Test Patterns book.

`XDocument.Parse` shouldn't be part of the _Act_ phase.

The test should be:

``` c#
[Theory, AutoData]
public void ScanCorrectlySerializesAttributedClassInstance(
    DataContractTestEventX @event)
{
    // Arrange
    // Act
    var actual = DataContractContentSerializer.Scan(@event.GetType().Assembly);
    var result = @event.AsSerializedString(actual);
    // Assert
    var expected = XDocument.Parse(
        "<test-event-x xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://grean.rocks/dc\">" +
        "  <number>" + @event.Number + "</number>" +
        "  <text>" + @event.Text + "</text>" +
        "</test-event-x>");
    Assert.Equal(expected, XDocument.Parse(result), new XNodeEqualityComparer());
}
```

ploeh replied:
> When the expected value requires some setup (in this example, string formatting, concatenation, etc) I usually put it in the Arrange phase. That might be a mistake on my part though, as I didn't look-up that in Gerard Meszaros's xUnit Test Patterns book.

It's not in [xUnit Test Patterns](http://bit.ly/xunitpatterns) - it's much more basic than that. In [Code Complete](http://amzn.to/1xioCmw) there's discussion about only adding the code when you need it. In this case, you don't _need_ it before the Assert phase.

Is `@event.AsSerializedString(actual)` part of the System Under Test?

moodmosaic replied:
Agreed about only adding the code when its needed. Again, up to now I've been making an exception when it comes to the `expected` value. Now it's totally clear though – Thank you for pointing to the direction that makes this explicit!

> Is @event.AsSerializedString(actual) part of the System Under Test?

It's not – but it's the only part of the test where we actually _do_ something with `actual`.

Is it part of the _Arrange_ phase? No. – Is it part of the _Assert_ phase? Probably not, because it's more related with the value associated with `actual`.

That's why `@event.AsSerializedString(actual)` exists in the _Act_ phase...

ploeh replied:
> It's not – but it's the only part of the test where we actually do something with actual.

By convention, we use the variable name `actual` for the result of exercising the SUT. Once we have an `actual` variable, we should be done with the _Act_ phase. The rest is verification.

moodmosaic replied:
> Once we have an actual variable, we should be done with the Act phase. The rest is verification.

I find that a valuable observation (like the [previous](https://github.com/GreanTech/AtomEventStore/pull/102/files#r19909874) one).

---

That means we can now re-write the test as:

``` c#
[Theory, AutoData]
public void ScanCorrectlySerializesAttributedClassInstance(
    DataContractTestEventX @event)
{
    var actual = DataContractContentSerializer.Scan(@event.GetType().Assembly);

   var expected = XDocument.Parse(
        "<test-event-x xmlns:i=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns=\"http://grean.rocks/dc\">" +
        "  <number>" + @event.Number + "</number>" +
        "  <text>" + @event.Text + "</text>" +
        "</test-event-x>");
    Assert.Equal(
        expected,
        XDocument.Parse(@event.AsSerializedString(actual),
        new XNodeEqualityComparer());
}
```

ploeh replied:
Looks good to me :+1:

moodmosaic replied:
Done in [58b9ddb](https://github.com/moodmosaic/AtomEventStore/commit/58b9ddb73c2a0bf9cb794e7e53837e8eee82a7db).

ploeh commented on AtomEventStore.UnitTests/DataContractContentSerializerTests.cs:
```diff
@@ -113,5 +113,61 @@ where t.IsDefined(a.GetType(), inherit: false)
+            DataContractTestEventX dctex)
+        {
+            var assemblyToScanForEvents = dctex.GetType().Assembly;
+            var sut =
```
Is this really the System Under Test?

moodmosaic replied:
I've attempted to address this in [d7b7ec7](https://github.com/moodmosaic/AtomEventStore/commit/d7b7ec78ae46893da2ff9f10e2b32711c3c796ce), by extrapolating from what we've discussed so far.

moodmosaic committed:
Refactored a test, according to the discussion at:
https://github.com/GreanTech/AtomEventStore/pull/102/files#r19240650.

The goal of this particular process is to produce the Scan method. The Scan
method is the System Under Test.

The Scan method returns an instance of DataContractContentSerializer and
the behaviour of that class doesn't change as a result of this test being
added.

To make it explicit, the System Under Test is the Scan method. That's the
method added after having written the test. There's no sut object, because
the SUT isn't an object, but a static method.

By convention, we use the variable name actual for the result of exercising
the SUT. Once we have an actual variable, we should be done with
the Act phase.

The rest is verification.

moodmosaic committed:
Refactored a test, according to the discussion at:
https://github.com/GreanTech/AtomEventStore/pull/102/files#r19240650.

The goal of this particular process is to produce the Scan method. The Scan
method is the System Under Test.

The Scan method returns an instance of DataContractContentSerializer and
the behaviour of that class doesn't change as a result of this test being
added.

To make it explicit, the System Under Test is the Scan method. That's the
method added after having written the test. There's no sut object, because
the SUT isn't an object, but a static method.

By convention, we use the variable name actual for the result of exercising
the SUT. Once we have an actual variable, we should be done with
the Act phase.

The rest is verification.

moodmosaic committed:
Constrained a code fragment to 80 chars width.

ploeh commented:
Thank you for your contribution! It's now live as [AtomEventStore 1.6.0](https://www.nuget.org/packages/AtomEventStore/1.6.0).

moodmosaic commented:
And thank _you_ for the great and diligent help on this pull request!
